### PR TITLE
[file-system][js] Make network task cancelable

### PR DESCRIPTION
**Pull request by lukmccall**
# Why

Added ability to cancel network tasks and a progress listener for file upload.

# How

- Created a `NetworkTask` class that is a basic class for download and upload tasks. 
- Added an upload progress listener. 

# TODO

- [x] iOS - https://github.com/expo/expo/pull/13320
- [x] Android - https://github.com/expo/expo/pull/13319
- [ ] Docs

# Test Plan

- tested using:
server:
```js
const express = require("express");
const app = express();
const fs = require("fs");
const multer = require("multer");
const upload = multer({ dest: "uploads/" });

// This method will save the binary content of the request as a file.
app.post("/binary-upload", (req, res) => {
  req.pipe(fs.createWriteStream("./uploads/image" + Date.now() + ".mp4"));
  res.end("OK");
});

// This method will save a "photo" field from the request as a file.
app.post("/multipart-upload", upload.single("photo"), (req, res) => {
  // You can access other HTTP parameters. They are located in the body object.
  console.log(req.body);
  res.end("OK");
});

app.listen(3000, () => {
  console.log("Working on port 3000");
});
```
test function:
```ts
<ListButton
          onPress={async () => {
            const [asset] = await Asset.loadAsync(require('../../assets/videos/t1.mp4'));
            // const [asset] = await Asset.loadAsync(require('../../assets/images/example1.jpg'));

            console.log({ asset });
            const task = FileSystem.createUploadTask(
              'http://192.168.0.136:3000/binary-upload',
              asset.localUri,
			 {
                uploadType: FileSystem.FileSystemUploadType.MULTIPART,
                fieldName: 'photo',
              },
              data => {
                console.log({ data });
              }
            );
            task.uploadAsync().then(x => {
              console.log({ x });
            });
            // await task.cancelAsync();
            // console.log('Cancel');
          }}
          title="test"
        />
```

### CODE REVIEW

File: packages/expo-file-system/src/FileSystem.ts [hunk 1]
```diff
@@ -271,49 +399,29 @@ export class DownloadResumable {
     if (!ExponentFileSystem.downloadResumableStartAsync) {
       throw new UnavailabilityError('expo-file-system', 'downloadResumableStartAsync');
     }
-    this._addSubscription();
+
+    if (this.checkIfTaskWasCanceled()) {
+      return undefined;
```
Comment (bbarthec) [inline packages/expo-file-system/src/FileSystem.ts:404]: That's the common way to return `undefined` 😉 
```suggestion
      return;
```

File: packages/expo-file-system/src/FileSystem.ts [hunk 2]
```diff
@@ -206,61 +207,188 @@ export function createDownloadResumable(
...
   }
 
   async downloadAsync(): Promise<FileSystemDownloadResult | undefined> {
     if (!ExponentFileSystem.downloadResumableStartAsync) {
       throw new UnavailabilityError('expo-file-system', 'downloadResumableStartAsync');
     }
-    this._addSubscription();
+
+    if (this.checkIfTaskWasCanceled()) {
+      return undefined;
```
Comment (bbarthec) [inline packages/expo-file-system/src/FileSystem.ts:362]:
```suggestion
      return;
```

File: packages/expo-file-system/src/FileSystem.ts [hunk 3]
```diff
@@ -206,61 +207,188 @@ export function createDownloadResumable(
...
+    return this.callback;
+  }
+
+  public async uploadAsync(): Promise<FileSystemUploadResult | undefined> {
+    if (!ExponentFileSystem.uploadTaskStartAsync) {
+      throw new UnavailabilityError('expo-file-system', 'uploadTaskStartAsync');
+    }
+
+    if (this.checkIfTaskWasCanceled()) {
+      return undefined;
```
Comment (bbarthec) [inline packages/expo-file-system/src/FileSystem.ts:317]:
```suggestion
      return;
```

File: packages/expo-file-system/src/FileSystem.ts [hunk 4]
```diff
@@ -206,61 +207,188 @@ export function createDownloadResumable(
...
+    if (this.subscription) {
+      return;
+    }
+
+    this.subscription = this.emitter.addListener(this.getEventName(), (event: ProgressEvent<T>) => {
+      if (event.uuid === this.uuid) {
+        const callback = this.getCallback();
+        if (callback) {
+          callback(event.data);
+        }
```
Comment (bbarthec) [inline packages/expo-file-system/src/FileSystem.ts:268]:
```suggestion
        callback?.(event.data);
```

File: packages/expo-file-system/src/FileSystem.ts [hunk 5]
```diff
@@ -206,61 +207,188 @@ export function createDownloadResumable(
...
+    if (!ExponentFileSystem.networkTaskCancelAsync) {
+      throw new UnavailabilityError('expo-file-system', 'networkTaskCancelAsync');
+    }
+
+    this.removeSubscription();
+    this.taskWasCanceled = true;
+    return await ExponentFileSystem.networkTaskCancelAsync(this.uuid);
+  }
+
+  protected checkIfTaskWasCanceled(): boolean {
```
Comment (bbarthec) [inline packages/expo-file-system/src/FileSystem.ts:241]:
```suggestion
  protected isTaskCancelled(): boolean {
```

File: packages/expo-file-system/src/FileSystem.ts [hunk 6]
```diff
@@ -206,61 +207,188 @@ export function createDownloadResumable(
...
+export function createUploadTask(
+  url: string,
+  fileUri: string,
+  options?: FileSystemUploadOptions,
+  callback?: NetworkTaskProgressCallback<UploadProgressData>
+): UploadTask {
+  return new UploadTask(url, fileUri, options, callback);
+}
+
+export abstract class NetworkTask<T extends DownloadProgressData | UploadProgressData> {
```
Comment (bbarthec) [inline packages/expo-file-system/src/FileSystem.ts:225]:
```suggestion
export abstract class CancellableNetworkTask<T extends DownloadProgressData | UploadProgressData> {
```

**Review by bbarthec — APPROVED** [review]
Looks fine to me 👍

File: packages/expo-file-system/src/FileSystem.ts [hunk 7]
```diff
@@ -238,7 +241,7 @@ export abstract class NetworkTask<T extends DownloadProgressData | UploadProgres
     return await ExponentFileSystem.networkTaskCancelAsync(this.uuid);
   }
 
-  protected checkIfTaskWasCanceled(): boolean {
+  protected isTaskCancelled(): boolean {
```
Comment (bbarthec) [inline packages/expo-file-system/src/FileSystem.ts:244]: I'm wondering about this change, as `isTaskCancelled` suggest it's a property of this class 🤔 
Maybe we should make it an actual property with `getter` and `setter`? 🤔

**Review by bbarthec — COMMENTED** [review]
This PR makes me wonder about proper namespacing of exported constants/types 🤔 
Should we flatten the namespace (module name) into the exported variable name or should we keep them namespaced? (`FileSystemVaribale` vs `FileSystem.Variable`) 🤔
cc @brentvatne, @ide

File: packages/expo-file-system/src/FileSystem.ts (hunk 7 above)
Comment (bbarthec) [reply packages/expo-file-system/src/FileSystem.ts:244]: I'm sorry that I'm changing my mind now 😅 
I've just realised it was me that suggested this change 😂

**Comment by developius** [comment]
@lukmccall @bbarthec I'm trying to make use of this `createUploadTask` method to show a progress indicator.
It looks like the events are being delivered out of order, but whether that's due to the underlying code or something related to this PR, I'm not sure.

This is the output I get from the callback:

```
{ "totalByteSent": 65535, "totalBytesExpectedToSend": 1766121 }
{ "totalByteSent": 983041, "totalBytesExpectedToSend": 1766121 }
{ "totalByteSent": 32768, "totalBytesExpectedToSend": 1766121 }
{ "totalByteSent": 491519, "totalBytesExpectedToSend": 1766121 }
{ "totalByteSent": 193258, "totalBytesExpectedToSend": 1766121 }
```

Have I done something wrong, or is there no guarantee on the delivery order of the events? And if the latter, what's a recommended way to deal with this - maybe only showing progress if the number increases?

**Comment by bbarthec** [comment]
@developius, do you mind creating an issue describing the problematic scenario you've reported? 
Looking at the message you've provided I have no basic information (affected platform(s), actual MVP/snack example) to decide whether you've done something wrong or right 🤷 Regarding order guarantee I delegate this question to @lukmccall 👀

**Comment by lukmccall** [comment]
Hi @developius, could you create a new issue with more information? Events should be delivered in order.

**Comment by developius** [comment]
@lukmccall @bbarthec sure, apologies for not including a reproduction. I'll prepare a new issue asap.

**Pull request merged**